### PR TITLE
Various fixes related to text

### PR DESCRIPTION
**Pull request by Happypig375**

### CODE REVIEW

**Comment by Happypig375** [comment]
`System.DllNotFoundException: Unable to load DLL 'libSkiaSharp': The specified module could not be found. (Exception from HRESULT: 0x8007007E)`??

**Comment by Happypig375** [comment]
Need help in diagnosing that error.

**Comment by wieslawsoltes** [comment]
> `System.DllNotFoundException: Unable to load DLL 'libSkiaSharp': The specified module could not be found. (Exception from HRESULT: 0x8007007E)`??

Which project did you run and on what platform? Might be missing native assets for that platform.

**Comment by Happypig375** [comment]
@wieslawsoltes It's from the CI.

File: src/Svg.Model/SvgExtensions.Painting.cs [hunk 1]
```diff
@@ -1004,7 +1004,8 @@ internal static bool IsValidStroke(SvgElement svgElement, SKRect skBounds)
         };
 
         var server = svgVisualElement.Fill;
-        var opacity = AdjustSvgOpacity(svgVisualElement.FillOpacity);
+        var opacity = AdjustSvgOpacity(svgVisualElement.Opacity);
+        opacity *= AdjustSvgOpacity(svgVisualElement.FillOpacity);
```
Comment (wieslawsoltes) [inline src/Svg.Model/SvgExtensions.Painting.cs:1008]: This breaks stuff

image

File: src/Svg.Model/SvgExtensions.Painting.cs [hunk 2]
```diff
@@ -1022,7 +1023,8 @@ internal static bool IsValidStroke(SvgElement svgElement, SKRect skBounds)
         };
 
         var server = svgVisualElement.Stroke;
-        var opacity = AdjustSvgOpacity(svgVisualElement.StrokeOpacity);
+        var opacity = AdjustSvgOpacity(svgVisualElement.Opacity);
+        opacity *= AdjustSvgOpacity(svgVisualElement.StrokeOpacity);
```
Comment (wieslawsoltes) [inline src/Svg.Model/SvgExtensions.Painting.cs:1027]: This breaks stuff

image

**Review by wieslawsoltes — CHANGES_REQUESTED** [review]
Please revert opacity changes from this pR

File: tests/Svg.Skia.UnitTests/SKSvgTests.cs [hunk 3]
```diff
@@ -1,5 +1,20 @@
...
+        static string CaseFile(string name) =>
+            Path.Combine("..", "..", "..", "Test cases", name);
+        var svg = new SKSvg();
+        using var _ = svg.Load(CaseFile("Sign in.svg"));
+        svg.Save(CaseFile("Sign in (Actual).png"), SkiaSharp.SKColors.Transparent);
+
+        using var expected = SkiaSharp.SKBitmap.Decode(CaseFile("Sign in.png"));
+        using var actual = SkiaSharp.SKBitmap.Decode(CaseFile("Sign in (Actual).png"));
+        Assert.Equal(expected.Info, actual.Info);
+        Assert.Equal(expected.Bytes, actual.Bytes);
```
Comment (wieslawsoltes) [inline tests/Svg.Skia.UnitTests/SKSvgTests.cs:18]: This will not work the same on all platforms, as rendering can use different rasterizers. Please revert that change or implement using similar technique as in original SVG repo,

**Review by wieslawsoltes — CHANGES_REQUESTED** [review]

**Comment by wieslawsoltes** [comment]
The issue with opacity is that text drawable does not use property Opacity property as others drawable do by creating opacity layer, this needs another way of fixing by refactoring TextDrawable, it's know issue, but refactoring is not trivial.

File: src/Svg.Model/Drawables/Elements/TextDrawable.cs [hunk 4]
```diff
@@ -110,7 +110,7 @@ internal string PrepareText(SvgTextBase svgTextBase, string value)
     {
         value = ApplyTransformation(svgTextBase, value);
         value = new StringBuilder(value).Replace("\r\n", " ").Replace('\r', ' ').Replace('\n', ' ').Replace('\t', ' ').ToString();
-        return svgTextBase.SpaceHandling == XmlSpaceHandling.Preserve ? value : s_multipleSpaces.Replace(value.Trim(), " ");
+        return svgTextBase.SpaceHandling == XmlSpaceHandling.Preserve ? value : s_multipleSpaces.Replace(value.TrimStart(), " ");
```
Comment (wieslawsoltes) [inline src/Svg.Model/Drawables/Elements/TextDrawable.cs:113]: Are you sure this is correct white space handling for all cases ?

File: src/Svg.Model/SvgExtensions.Painting.cs (hunk 2 above)
Comment (Happypig375) [reply src/Svg.Model/SvgExtensions.Painting.cs]: How do I view the same window?

File: src/Svg.Model/Drawables/Elements/TextDrawable.cs (hunk 4 above)
Comment (Happypig375) [reply src/Svg.Model/Drawables/Elements/TextDrawable.cs:113]: I compared with how Chrome renders

Comment (Happypig375) [reply src/Svg.Model/Drawables/Elements/TextDrawable.cs:113]: using that SVG I provided

File: tests/Svg.Skia.UnitTests/SKSvgTests.cs (hunk 3 above)
Comment (Happypig375) [reply tests/Svg.Skia.UnitTests/SKSvgTests.cs:18]: That app as shown above could be used instead I think.

File: src/Svg.Model/SvgExtensions.Painting.cs (hunk 2 above)
Comment (wieslawsoltes) [reply src/Svg.Model/SvgExtensions.Painting.cs]: Run https://github.com/wieslawsoltes/Svg.Skia/tree/master/samples/SvgToPng (Windows only)

**Comment by wieslawsoltes** [comment]
Those are test files that need to render correctly according to Svg standard.

Svg files
https://github.com/wieslawsoltes/SVG/tree/25b3700d88f7d39d146483cd00fb8eb4b582e5c7/Tests/W3CTestSuite/svg

Expected results
https://github.com/wieslawsoltes/SVG/tree/25b3700d88f7d39d146483cd00fb8eb4b582e5c7/Tests/W3CTestSuite/png

Test app
https://github.com/wieslawsoltes/Svg.Skia/tree/master/samples/SvgToPng?rgh-link-date=2022-08-24T08%3A01%3A14Z

File: src/Svg.Model/SvgExtensions.Painting.cs (hunk 2 above)
Comment (Happypig375) [reply src/Svg.Model/SvgExtensions.Painting.cs]: @wieslawsoltes It initializes empty. It seems to need to load a JSON file somewhere using "Load..."?

Comment (wieslawsoltes) [reply src/Svg.Model/SvgExtensions.Painting.cs]: @Happypig375 
https://user-images.githubusercontent.com/2297442/186384239-425c79b3-600f-4d82-8dee-c0d06ca87821.mp4

**Comment by Happypig375** [comment]
Great job CI, failing to write the failure

File: samples/svgc/Program.cs [hunk 5]
```diff
@@ -42,6 +43,14 @@ public SMP.SKImage LoadImage(Stream stream)
             Height = image.Height
         };
     }
+
+    public List<(string text, float advance, SMP.SKTypeface? typeface)>
+        FindTypefaces(string text, SMP.SKPaint paintPreferredTypeface)
+    {
+        // TODO: How can font fallback be done in a code generator?
```
Comment (wieslawsoltes) [inline samples/svgc/Program.cs:50]: That is issue with generator not working with SkiaSharp native assets for platforms, so probably not possible or very hard requiring some package reference hacks.

File: src/Avalonia.Svg/AvaloniaModelExtensions.cs [hunk 6]
```diff
@@ -117,61 +117,50 @@ public static AM.TextAlignment ToTextAlignment(this SKTextAlign textAlign)
...
-                return AM.FontStyle.Normal;
-            case SKFontStyleSlant.Italic:
-                return AM.FontStyle.Italic;
-
-            case SKFontStyleSlant.Oblique:
-                return AM.FontStyle.Oblique;
-        }
-    }
+    public static AM.FontWeight ToFontWeight(this SKFontStyleWeight fontStyleWeight) =>
+        // The numeric values in these two enumerations have the same scale.
```
Comment (wieslawsoltes) [inline src/Avalonia.Svg/AvaloniaModelExtensions.cs:121]: Please do not do unrelated refactoring in this PR, please keep changes simple to one feature otherwise reviewing is very hard and time consuming. Best to revert all unrelated changes in this PR.

File: src/Avalonia.Svg/AvaloniaModelExtensions.cs [hunk 7]
```diff
@@ -117,61 +117,50 @@ public static AM.TextAlignment ToTextAlignment(this SKTextAlign textAlign)
...
+        // Casting is even encouraged in AM.FontWeight XML documentation.
+        (AM.FontWeight)fontStyleWeight;
+    public static SKFontStyleWeight ToSKFontWeight(this AM.FontWeight fontStyleWeight) =>
+        (SKFontStyleWeight)fontStyleWeight;
+
+    public static AM.FontStretch ToFontStretch(this SKFontStyleWidth fontStyleWidth) => fontStyleWidth switch {
```
Comment (wieslawsoltes) [inline src/Avalonia.Svg/AvaloniaModelExtensions.cs:127]: Please do not do such refactoring in this PR

File: src/Avalonia.Svg/AvaloniaModelExtensions.cs [hunk 8]
```diff
@@ -117,61 +117,50 @@ public static AM.TextAlignment ToTextAlignment(this SKTextAlign textAlign)
...
+        SKFontStyleWidth.Condensed => AM.FontStretch.Condensed,
+        SKFontStyleWidth.SemiCondensed => AM.FontStretch.SemiCondensed,
+        SKFontStyleWidth.Normal => AM.FontStretch.Normal,
+        SKFontStyleWidth.SemiExpanded => AM.FontStretch.SemiExpanded,
+        SKFontStyleWidth.Expanded => AM.FontStretch.Expanded,
+        SKFontStyleWidth.ExtraExpanded => AM.FontStretch.ExtraExpanded,
+        SKFontStyleWidth.UltraExpanded => AM.FontStretch.UltraExpanded,
+        _ => AM.FontStretch.Normal
+    };
+    public static SKFontStyleWidth ToSKFontStretch(this AM.FontStretch fontStyleWidth) => fontStyleWidth switch {
```
Comment (wieslawsoltes) [inline src/Avalonia.Svg/AvaloniaModelExtensions.cs:139]: Please do not do such refactoring in this PR

File: src/Avalonia.Svg/AvaloniaModelExtensions.cs [hunk 9]
```diff
@@ -117,61 +117,50 @@ public static AM.TextAlignment ToTextAlignment(this SKTextAlign textAlign)
...
+        AM.FontStretch.SemiCondensed => SKFontStyleWidth.SemiCondensed,
+        AM.FontStretch.Normal => SKFontStyleWidth.Normal,
+        AM.FontStretch.SemiExpanded => SKFontStyleWidth.SemiExpanded,
+        AM.FontStretch.Expanded => SKFontStyleWidth.Expanded,
+        AM.FontStretch.ExtraExpanded => SKFontStyleWidth.ExtraExpanded,
+        AM.FontStretch.UltraExpanded => SKFontStyleWidth.UltraExpanded,
+        _ => SKFontStyleWidth.Normal
+    };
+
+    public static AM.FontStyle ToFontStyle(this SKFontStyleSlant fontStyleSlant) => fontStyleSlant switch
```
Comment (wieslawsoltes) [inline src/Avalonia.Svg/AvaloniaModelExtensions.cs:152]: Please do not do such refactoring in this PR

File: src/Avalonia.Svg/AvaloniaModelExtensions.cs [hunk 10]
```diff
@@ -117,61 +117,50 @@ public static AM.TextAlignment ToTextAlignment(this SKTextAlign textAlign)
...
+    {
+        SKFontStyleSlant.Italic => AM.FontStyle.Italic,
+        SKFontStyleSlant.Oblique => AM.FontStyle.Oblique,
+        _ => AM.FontStyle.Normal
+    };
+    public static SKFontStyleSlant ToSKFontStyle(this AM.FontStyle fontStyleSlant) => fontStyleSlant switch
```
Comment (wieslawsoltes) [inline src/Avalonia.Svg/AvaloniaModelExtensions.cs:158]: Please do not do such refactoring in this PR

File: src/ShimSkiaSharp/SKPaint.cs [hunk 11]
```diff
@@ -1,6 +1,6 @@
 ﻿namespace ShimSkiaSharp;
 
-public sealed class SKPaint
+public sealed record class SKPaint
```
Comment (wieslawsoltes) [inline src/ShimSkiaSharp/SKPaint.cs:3]: Please revert, it should match original SkiaSharp type.

File: src/ShimSkiaSharp/SKTypeface.cs [hunk 12]
```diff
@@ -1,6 +1,6 @@
 ﻿namespace ShimSkiaSharp;
 
-public class SKTypeface
+public record class SKTypeface
```
Comment (wieslawsoltes) [inline src/ShimSkiaSharp/SKTypeface.cs:3]: Please revert, it should match original SkiaSharp type.

File: src/Svg.Model/Drawables/Elements/TextDrawable.cs [hunk 13]
```diff
@@ -256,138 +257,192 @@ internal void EndDraw(SKCanvas skCanvas, DrawAttributes ignoreAttributes, MaskDr
...
-        GetPositionsDX(svgTextBase, skViewport, dxs);
-        GetPositionsDY(svgTextBase, skViewport, dys);
-
-        if (xs.Count >= 1 && ys.Count >= 1 && xs.Count == ys.Count && xs.Count == text.Length)
-        {
-            // TODO: Fix text position rendering.
-            var points = new SKPoint[xs.Count];
-
-            for (var i = 0; i < xs.Count; i++)
+        foreach (var node in GetContentNodes(svgTextBase))
```
Comment (wieslawsoltes) [inline src/Svg.Model/Drawables/Elements/TextDrawable.cs:323]: Please add braces, please add everywhere else where code was changed .

File: src/Svg.Model/IAssetLoader.cs [hunk 14]
```diff
@@ -10,4 +11,6 @@ namespace Svg.Model;
 public interface IAssetLoader
 {
     SKImage LoadImage(Stream stream);
+    List<(string text, float advance, SKTypeface? typeface)>
```
Comment (wieslawsoltes) [inline src/Svg.Model/IAssetLoader.cs:14]: Please don't do such log tuple types, just add simple class or record, it's hard to read.

File: tests/Tests using SvgToPng/_SvgToPng.json [hunk 15]
```diff
@@ -0,0 +1,8 @@
+[
```
Comment (wieslawsoltes) [inline tests/Tests using SvgToPng/_SvgToPng.json:1]: Remove this file

File: src/Avalonia.Svg/AvaloniaModelExtensions.cs (hunk 9 above)
Comment (Happypig375) [reply src/Avalonia.Svg/AvaloniaModelExtensions.cs:152]: Required as part of AvaloniaAssetLoader implementation

File: src/Avalonia.Svg/AvaloniaModelExtensions.cs (hunk 8 above)
Comment (Happypig375) [reply src/Avalonia.Svg/AvaloniaModelExtensions.cs:139]: This is part of AvaloniaAssetLoader implementation

File: src/Avalonia.Svg/AvaloniaModelExtensions.cs (hunk 10 above)
Comment (Happypig375) [reply src/Avalonia.Svg/AvaloniaModelExtensions.cs:158]: This is part of AvaloniaAssetLoader implementation

**Comment by wieslawsoltes** [comment]
@Happypig375 The real issue with opacity is when used in combination with filter effects, below is minimal repro:

```svg
<svg xmlns="http://www.w3.org/2000/svg" width="390" height="844" viewBox="0 0 390 844">

  <defs>

    <filter id="Path_129" x="44" y="602" width="302" height="53" filterUnits="userSpaceOnUse">
      <feOffset dy="6" input="SourceAlpha"/>
      <feGaussianBlur stdDeviation="3" result="blur-3"/>
      <feFlood flood-opacity="0.161"/>
      <feComposite operator="in" in2="blur-3"/>
      <feComposite in="SourceGraphic"/>
    </filter>

  </defs>
  
    <g filter="url(#Path_129)">
      <path id="Path_129-2" d="M17.5,0h249a17.5,17.5,0,0,1,0,35H17.5a17.5,17.5,0,0,1,0-35Z" transform="translate(53 605)" fill="#454984" opacity="0.282"/>
    </g>

</svg>
```

Sign in - filter opacity issue

**Comment by Happypig375** [comment]
@wieslawsoltes Svg.Skia ignores the opacity attribute in code when getting the source graphic for filter effects. I think this is in line with spec: https://www.w3.org/TR/SVG11/filters.html#AccessingBackgroundImage
Where major browsers (surprisingly excluding Internet Explorer) do not respect it:
Image linked in spec - image
Edge (Chromium) - image
Firefox (Gecko) - 
image
Netscape (Yes, the unmaintained one from decades ago) - 
image
Internet Explorer (Trident) -
image

Meanwhile, all browsers agree that the opacity applied on that path is respected instead of ignored.

Is this a case where all major implementers agree to deviate from the spec while we followed?

**Comment by Happypig375** [comment]
I'll split that into another PR

**Comment by wieslawsoltes** [comment]
@Happypig375 Added a way to compare images in unit tests
